Cabal/Distribution/Types/VersionRange.hs

```diff
@@ -340,34 +358,94 @@ withinRange v = foldVersionRange
 -- Wildcard range utilities
 --
 
--- | @since 2.2
-wildcardUpperBound :: Version -> Version
-wildcardUpperBound = alterVersion $
-    \lowerBound -> init lowerBound ++ [last lowerBound + 1]
-
 isWildcardRange :: Version -> Version -> Bool
 isWildcardRange ver1 ver2 = check (versionNumbers ver1) (versionNumbers ver2)
   where check (n:[]) (m:[]) | n+1 == m = True
         check (n:ns) (m:ms) | n   == m = check ns ms
         check _      _                 = False
 
--- | Compute next greater major version to be used as upper bound
---
--- Example: @0.4.1@ produces the version @0.5@ which then can be used
--- to construct a range @>= 0.4.1 && < 0.5@
---
--- @since 2.2
-majorUpperBound :: Version -> Version
-majorUpperBound = alterVersion $ \numbers -> case numbers of
-    []        -> [0,1] -- should not happen
-    [m1]      -> [m1,1] -- e.g. version '1'
-    (m1:m2:_) -> [m1,m2+1]
-
 -------------------------------------------------------------------------------
 -- Parsec & Pretty
 -------------------------------------------------------------------------------
 
-instance Pretty VersionRange where
+class Ord v => IsVersion v where
+  dispWild :: v -> Disp.Doc
+  noVersion :: VersionRangeOver v
+  majorUpperBound :: v -> v
+  wildcardUpperBound :: v -> v
+  verOrWild :: CabalParsing m => m (Bool, v)
```

Comment:
this is ugly.

Reply:
... I mean, we probably need an extension point to allow `version` specific `VersionRange` syntax constructs. And this isn't the prettiest one. Have to think what would work best (given `Version` and `PkgConfigVersion` differences).

IMO `openssl ==1.0.1*` isn't something we want to allow. I don't know what `pkg-configs` own syntax looks like.

Reply:
One idea is to bundle `versionRangeReadP` as a whole into `IsVersion`

Reply:
This is indeed the part I'm most unsure/unhappy about.